modules/rostests/kmtests/ntos_ob/ObInsert.c

```diff
@@ -0,0 +1,651 @@
+    RtlZeroMemory(Object, sizeof(DUMMY_TYPE));
+    Object->Data = 123;
+
+    ObjectHeader = OBJECT_TO_OBJECT_HEADER(Object);
+
+    ok_eq_ulong(ObjectHeader->PointerCount, 1LU);
+    ok_eq_ulong(ObjectHeader->HandleCount, 0LU);
+
+    /* Insert the object */
+    Status = ObInsertObject(Object,
+                            NULL,
+                            STANDARD_RIGHTS_ALL,
+                            0,
+                            NULL,
+                            &Handle);
+
+    ok_eq_hex(Status, STATUS_SUCCESS);
+    ok(Object != NULL, "Object is NULL\n");
+    ok(Handle != NULL, "Handle is NULL\n");
+
+    ok_eq_ulong(ObjectHeader->PointerCount, 2LU);
+    ok_eq_ulong(ObjectHeader->HandleCount, 1LU);
+
+    ObDereferenceObject(Object);
+
+    Status = ZwClose(Handle);
+    ok_eq_hex(Status, STATUS_SUCCESS);
+}
+
+/*!
+ * @brief This function tests STATUS_OBJECT_PATH_NOT_FOUND.
+ */
+static
+VOID
+ObInsert_PathNotFound()
+{
+    NTSTATUS Status;
+    UNICODE_STRING ObjectName;
+    PDUMMY_TYPE Object;
+    HANDLE Handle;
+    POBJECT_HEADER ObjectHeader;
+
+    /* Initialize the object name. To simulate STATUS_OBJECT_PATH_NOT_FOUND, provide
+       an invalid non-last path component ("ObInsertTestt") */
+    RtlInitUnicodeString(&ObjectName,
+                         L"\\ObInsertTestt\\DummyObject_PathNotFound");
+
+    /* Initialize object attributes with the name */
+    InitializeObjectAttributes(&DummyObjectAttributes,
+                               &ObjectName,
+                               OBJ_CASE_INSENSITIVE,
+                               NULL,
+                               NULL);
+
+    /* Create the object */
+    Status = ObCreateObject(KernelMode,
+                            DummyType,
+                            &DummyObjectAttributes,
+                            KernelMode,
+                            NULL,
+                            sizeof(DUMMY_TYPE),
+                            0,
+                            0,
+                            (PVOID *)&Object);
+
+    ok_eq_hex(Status, STATUS_SUCCESS);
+
+    /* Initialize the object data */
+    RtlZeroMemory(Object, sizeof(DUMMY_TYPE));
+    Object->Data = 123;
+
+    ObjectHeader = OBJECT_TO_OBJECT_HEADER(Object);
+
+    /* Check the object handles before insertion */
+    ok_eq_ulong(ObjectHeader->PointerCount, 1LU);
+    ok_eq_ulong(ObjectHeader->HandleCount, 0LU);
+
+    /* Insert the object */
+    Status = ObInsertObject(Object,
+                            NULL,
+                            STANDARD_RIGHTS_ALL,
+                            0,
+                            NULL,
+                            &Handle);
+
+    /* STATUS_OBJECT_PATH_NOT_FOUND because "ObInsertTestt" doesn't exist, but there are
+       remaining components of the path */
+    ok_eq_hex(Status, STATUS_OBJECT_PATH_NOT_FOUND);
+
+    /* Check the object handles after invalid insertion. ObInsertObject should've
+       automatically dereferenced the object */
```

Comment:
But if ObInsertObject has dereferenced the object and there are indeed no references to it anymore, there is a "risk" of said object to have been completely destroyed by now, and you doing dereferencing on invalid memory below.

Reply:
Thanks. Removed potentially unsafe object header checks after dereferencing. Instead, a delete callback procedure with an indicator like a counter can be used to check: https://github.com/reactos/reactos/pull/6968/commits/60b55df68420d4caeab7a8b4e695840183eedba5 The test checks if the deletion counter has increased after `ObInsertObject` fails.